kcp/controllers/watcher_controller.go

```diff
@@ -74,3 +114,312 @@ func (r *WatcherReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		}
+	}
+	return nil
+}
+
+func (r *WatcherReconciler) updateWatcherCRStatus(ctx context.Context, obj *componentv1alpha1.Watcher,
+	state componentv1alpha1.WatcherState, msg string) error {
+	obj.Status.State = state
+	switch state { //nolint:exhaustive
+	case componentv1alpha1.WatcherStateReady:
+		util.AddReadyCondition(obj, componentv1alpha1.ConditionStatusTrue, msg)
+	case "":
+		util.AddReadyCondition(obj, componentv1alpha1.ConditionStatusUnknown, msg)
+	default:
+		util.AddReadyCondition(obj, componentv1alpha1.ConditionStatusFalse, msg)
+	}
+	return r.Status().Update(ctx, obj.SetObservedGeneration())
+}
+
+func (r *WatcherReconciler) HandleDeletingState(ctx context.Context, logger logr.Logger,
+	obj *componentv1alpha1.Watcher) error {
+	err := r.deleteConfigMapForCR(ctx, obj)
+	if err != nil {
+		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to create config map")
+	}
+	err = r.deleteServiceMeshConfigForCR(ctx, obj)
+	if err != nil {
+		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to create or update service mesh config")
+	}
+	err = r.deleteSKRWatcherConfigForCR(ctx, obj)
+	if err != nil {
+		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to update SKR config")
+	}
+	updated := controllerutil.RemoveFinalizer(obj, watcherFinalizer)
+	if !updated {
+		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to remove finalizer")
+	}
+	err = r.Update(ctx, obj)
+	if err != nil {
+		logger.Error(err, "failed to update watcher cr")
+	}
+	logger.Info("deletion state handling was successful")
+	return nil
+}
+
+func (r *WatcherReconciler) deleteSKRWatcherConfigForCR(ctx context.Context, obj *componentv1alpha1.Watcher) error {
+	//this will be implemented as part of another step: see https://github.com/kyma-project/kyma-watcher/issues/33
+	return nil
+}
+
+func (r *WatcherReconciler) deleteServiceMeshConfigForCR(ctx context.Context, obj *componentv1alpha1.Watcher) error {
+	namespace := obj.GetNamespace()
+	vsName := obj.GetName()
+	istioClientSet, err := istioclient.NewForConfig(r.RestConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create istio client set from rest config(%s): %w", r.RestConfig.String(), err)
+	}
+	_, err = istioClientSet.NetworkingV1beta1().VirtualServices(namespace).Get(ctx, vsName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("failed to get istio virtual service: %w", err)
+	}
+	if errors.IsNotFound(err) {
+		//nothing to do
+		return nil
+	}
+	err = istioClientSet.NetworkingV1beta1().VirtualServices(namespace).Delete(ctx, vsName, metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to delete istio virtual service: %w", err)
+	}
+	return nil
+
+}
+
+func (r *WatcherReconciler) deleteConfigMapForCR(ctx context.Context, obj *componentv1alpha1.Watcher) error {
+	watcherCRLabels := obj.GetLabels()
+	_, ok := watcherCRLabels[defaultOperatorWatcherCRLabel]
```

Comment:
Please also check if the value is equal to `true` - Then it is easier for developer to deactivate, by just setting it to false, instead of deleting the label. Thus, both ways work.